"Possible precedence problem between ! and numeric eq (==)" failure in automake

<!--
Note: you can also replace the whole content with a file generated by the perlbug utility;
perlbug reports by email are no longer supported.
Be sure to enclose your perl configuration in a fenced code block to preserve formatting:
https://help.github.com/en/github/writing-on-github/creating-and-highlighting-code-blocks
-->

<!--
If your bug is about a Perl core module rather than a core language
feature, please enter its name after.
-->
Module:

**Description**
<!-- A clear and concise description of what the bug is. -->
When using Perl 5.41.7 or Perl 5.41.8, we are seeing this error message appear when running automake (written in perl).
Perl 5.41.3 did not have this problem.
```
Possible precedence problem between ! and numeric eq (==) at /data/zopen/usr/local/zopen/automake/automake-1.17.20241029_040611.zos/bin/automake line 6875.
```
Is this expected? It's preventing us from building autotools based projects.

Here's the reduced snippet of what automake is doing:
```
# test.pl
use strict;
use warnings;

sub transform_token {
    my ($token, $transform_ref, $key) = @_;
    return exists $transform_ref->{$key} ? $transform_ref->{$key} : '';
}

sub transform {
    my ($token, $transform) = @_;

    if ($token =~ /^%([\w\-]+)%$/) {
        return transform_token($token, $transform, $1);
    }
    elsif ($token =~ /^%\?([\w\-]+)%$/) {
        return transform_token($token, $transform, $1) ? 'TRUE' : 'FALSE';
    }
    elsif ($token =~ /^ \? (!?) ([\w\-]+) \? $/x) {
        my $neg = ($1 eq '!') ? 1 : 0;
        my $val = transform_token($token, $transform, $2);
        return (!!$val == $neg) ? '##%' : '';
    }
    else {
        die "Unknown request format: $token";
    }
}

# Simple test cases
my %transform_data = (
    'KEY1' => 'VALUE1',
    'KEY2' => '',
    'KEY3' => '1',
);

print transform('?KEY3?', \%transform_data) eq '##%' ? "PASS\n" : "FAIL\n";
```

**Steps to Reproduce**
<!-- A one-liner or script to reproduce the issue. -->
```
perl test.pl
```

**Expected behavior**
<!-- A clear and concise description of what you expected to happen. -->
No "Possible precedence problem between ! and numeric eq (==)" message

**Perl configuration**
<!-- Please paste `perl -V` output just below. -->
```
# perl -V output goes here
 perl -V
Summary of my perl5 (revision 5 version 41 subversion 7) configuration:
  Derived from: 3a3761b09845d0d6d68f91eb45c52ae70de89489
  Platform:
    osname=os390
    osvers=27.00
    archname=os390
    uname='os390 vs01 27.00 04 8562 '
    config_args='-Dprefix=/home/itodoro/zopen/usr/local/zopen/perl5/perl5 -Duserelocatableinc -Dusedevel -des -Duse64bitall -Dusedl -D ccflags=-fzos-le-char-mode=ascii -std=gnu11 -mnocsect -fno-short-enums -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -m64 -mzos-target=zosv2r4 -O3  -I/home/itodoro/zopen/usr/local/zopen/curl/curl/include -I/home/itodoro/zopen/usr/local/zopen/ncurses/ncurses/include -I/home/itodoro/zopen/usr/local/zopen/zlib/zlib/include -isystem /home/itodoro/zopen/usr/local/zopen/zoslib/zoslib/include -include /home/itodoro/zopen/usr/local/zopen/zoslib/zoslib/include/zos-v2r5-symbolfixes.h  -DNSIG=42 -D_XOPEN_SOURCE=600 -D_ALL_SOURCE -D_OPEN_SYS_FILE_EXT=1 -D_AE_BIMODAL=1 -D_ENHANCED_ASCII_EXT=0xFFFFFFFF  -DZOSLIB_OVERRIDE_CLIB=1 -DZOSLIB_OVERRIDE_CLIB_GETENV=1 -DOS390 -DZOS -DNO_LOCALE_MESSAGES -DOEMVS -DMAXSIG=39 -D_EXT=1 -fvisibility=default -D ldflags=-Wl,-bedit=no -m64  -L/home/itodoro/zopen/usr/local/zopen/curl/curl/lib -L/home/itodoro/zopen/usr/local/zopen/ncurses/ncurses/lib -L/home/itodoro/zopen/usr/local/zopen/zlib/zlib/lib -L/home/itodoro/zopen/usr/local/zopen/zoslib/zoslib/lib -D libs=  -lcurl -lncurses -lz -lzoslib /home/itodoro/zopen/usr/local/zopen/zoslib/zoslib/lib/celquopt.s.o -lzoslib-supp /jenkins/workspace/Port-Build/perl/perl5/.zoslib_hooks/zoslib_env_hook.c.o CC=clang CPPFLAGS=-DNSIG=42 -D_XOPEN_SOURCE=600 -D_ALL_SOURCE -D_OPEN_SYS_FILE_EXT=1 -D_AE_BIMODAL=1 -D_ENHANCED_ASCII_EXT=0xFFFFFFFF  -DZOSLIB_OVERRIDE_CLIB=1 -DZOSLIB_OVERRIDE_CLIB_GETENV=1 CFLAGS=-fzos-le-char-mode=ascii -std=gnu11 -mnocsect -fno-short-enums -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -m64 -mzos-target=zosv2r4 -O3  -I/home/itodoro/zopen/usr/local/zopen/curl/curl/include -I/home/itodoro/zopen/usr/local/zopen/ncurses/ncurses/include -I/home/itodoro/zopen/usr/local/zopen/zlib/zlib/include -isystem /home/itodoro/zopen/usr/local/zopen/zoslib/zoslib/include -include /home/itodoro/zopen/usr/local/zopen/zoslib/zoslib/include/zos-v2r5-symbolfixes.h  CXX=clang++ CXXFLAGS=-fzos-le-char-mode=ascii -mnocsect -fno-short-enums -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -m64 -mzos-target=zosv2r4 -O3     -I/home/itodoro/zopen/usr/local/zopen/curl/curl/include -I/home/itodoro/zopen/usr/local/zopen/ncurses/ncurses/include -I/home/itodoro/zopen/usr/local/zopen/ncurses/ncurses/include/ncurses -I/home/itodoro/zopen/usr/local/zopen/zlib/zlib/include -isystem /home/itodoro/zopen/usr/local/zopen/zoslib/zoslib/include -include /home/itodoro/zopen/usr/local/zopen/zoslib/zoslib/include/zos-v2r5-symbolfixes.h  LDFLAGS=-Wl,-bedit=no -m64  -L/home/itodoro/zopen/usr/local/zopen/curl/curl/lib -L/home/itodoro/zopen/usr/local/zopen/ncurses/ncurses/lib -L/home/itodoro/zopen/usr/local/zopen/zlib/zlib/lib -L/home/itodoro/zopen/usr/local/zopen/zoslib/zoslib/lib LIBS=  -lcurl -lncurses -lz -lzoslib /home/itodoro/zopen/usr/local/zopen/zoslib/zoslib/lib/celquopt.s.o -lzoslib-supp /jenkins/workspace/Port-Build/perl/perl5/.zoslib_hooks/zoslib_env_hook.c.o LDLIBS=  -lcurl -lncurses -lz -lzoslib /home/itodoro/zopen/usr/local/zopen/zoslib/zoslib/lib/celquopt.s.o -lzoslib-supp /jenkins/workspace/Port-Build/perl/perl5/.zoslib_hooks/zoslib_env_hook.c.o'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=undef
    usemultiplicity=undef
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
  Compiler:
    cc='clang'
    ccflags ='-fzos-le-char-mode=ascii -std=gnu11 -mnocsect -fno-short-enums -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -m64 -mzos-target=zosv2r4 -O3 -I/home/itodoro/zopen/usr/local/zopen/curl/curl/include -I/home/itodoro/zopen/usr/local/zopen/ncurses/ncurses/include -I/home/itodoro/zopen/usr/local/zopen/zlib/zlib/include -isystem /home/itodoro/zopen/usr/local/zopen/zoslib/zoslib/include -include /home/itodoro/zopen/usr/local/zopen/zoslib/zoslib/include/zos-v2r5-symbolfixes.h -DNSIG=42 -D_XOPEN_SOURCE=600 -D_ALL_SOURCE -D_OPEN_SYS_FILE_EXT=1 -D_AE_BIMODAL=1 -D_ENHANCED_ASCII_EXT=0xFFFFFFFF -DZOSLIB_OVERRIDE_CLIB=1 -DZOSLIB_OVERRIDE_CLIB_GETENV=1 -DOS390 -DZOS -DNO_LOCALE_MESSAGES -DOEMVS -DMAXSIG=39 -D_EXT=1 -fvisibility=default'
    optimize=' '
    cppflags='-fzos-le-char-mode=ascii -std=gnu11 -mnocsect -fno-short-enums -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -m64 -mzos-target=zosv2r4 -O3 -I/home/itodoro/zopen/usr/local/zopen/curl/curl/include -I/home/itodoro/zopen/usr/local/zopen/ncurses/ncurses/include -I/home/itodoro/zopen/usr/local/zopen/zlib/zlib/include -isystem /home/itodoro/zopen/usr/local/zopen/zoslib/zoslib/include -include /home/itodoro/zopen/usr/local/zopen/zoslib/zoslib/include/zos-v2r5-symbolfixes.h -DNSIG=42 -D_XOPEN_SOURCE=600 -D_ALL_SOURCE -D_OPEN_SYS_FILE_EXT=1 -D_AE_BIMODAL=1 -D_ENHANCED_ASCII_EXT=0xFFFFFFFF -DZOSLIB_OVERRIDE_CLIB=1 -DZOSLIB_OVERRIDE_CLIB_GETENV=1 -DOS390 -DZOS -DNO_LOCALE_MESSAGES -DOEMVS -DMAXSIG=39 -D_EXT=1 -fvisibility=default'
    ccversion=''
    gccversion=''
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=87654321
    doublekind=4
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=2
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='clang'
    ldflags ='-Wl,-bedit=no -m64 -L/home/itodoro/zopen/usr/local/zopen/curl/curl/lib -L/home/itodoro/zopen/usr/local/zopen/ncurses/ncurses/lib -L/home/itodoro/zopen/usr/local/zopen/zlib/zlib/lib -L/home/itodoro/zopen/usr/local/zopen/zoslib/zoslib/lib'
    libpth=/lib /usr/lib
    libs= -lcurl -lncurses -lz -lzoslib /home/itodoro/zopen/usr/local/zopen/zoslib/zoslib/lib/celquopt.s.o -lzoslib-supp /jenkins/workspace/Port-Build/perl/perl5/.zoslib_hooks/zoslib_env_hook.c.o
    perllibs=-lcurl -lncurses -lz -lzoslib /home/itodoro/zopen/usr/local/zopen/zoslib/zoslib/lib/celquopt.s.o -lzoslib-supp /jenkins/workspace/Port-Build/perl/perl5/.zoslib_hooks/zoslib_env_hook.c.o
    libc=
    so=so
    useshrplib=true
    libperl=libperl.so
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dllload.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='  /jenkins/workspace/Port-Build/perl/perl5/libperl.x'
    cccdlflags='-c -m64'
    lddlflags='-m64 -L/home/itodoro/zopen/usr/local/zopen/curl/curl/lib -L/home/itodoro/zopen/usr/local/zopen/ncurses/ncurses/lib -L/home/itodoro/zopen/usr/local/zopen/zlib/zlib/lib -L/home/itodoro/zopen/usr/local/zopen/zoslib/zoslib/lib /jenkins/workspace/Port-Build/perl/perl5/libperl.x'


Characteristics of this binary (from libperl):
  Compile-time options:
    HAS_LONG_DOUBLE
    HAS_STRTOLD
    HAS_TIMES
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_HASH_FUNC_SIPHASH13
    PERL_HASH_USE_SBOX32
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    PERL_USE_DEVEL
    PERL_USE_SAFE_PUTENV
    USE_64_BIT_ALL
    USE_64_BIT_INT
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
  Locally applied patches:
    uncommitted-changes
  Built under os390
  Compiled at Jan 23 2025 11:25:38
  %ENV:
    PERL5LIB="/home/itodoro/zopen/usr/local/zopen/perl/perl5-heads.v5.41.7.20250123_102326.zos/lib/5.41.7/:/home/itodoro/zopen/usr/local/zopen/perl/perl5-heads.v5.41.7.20250123_102326.zos/lib/5.41.7/os390/:/home/itodoro/zopen/usr/local/zopen/perl/perl5-heads.v5.41.7.20250123_102326.zos/lib/perl5/5.41.7/:/home/itodoro/zopen/usr/local/zopen/perl/perl5-heads.v5.41.7.20250123_102326.zos/lib/perl5/5.41.7/os390/:/home/itodoro/zopen/usr/local/zopen/perl/perl5-heads.v5.41.7.20250123_102326.zos/lib/5.41.7/:/home/itodoro/zopen/usr/local/zopen/perl/perl5-heads.v5.41.7.20250123_102326.zos/lib/5.41.7/os390/:/home/itodoro/zopen/usr/local/zopen/perl/perl5-heads.v5.41.7.20250123_102326.zos/lib/perl5/5.41.7/:/home/itodoro/zopen/usr/local/zopen/perl/perl5-heads.v5.41.7.20250123_102326.zos/lib/perl5/5.41.7/os390/:"
  @INC:
    /home/itodoro/zopen/usr/local/zopen/perl/perl5-heads.v5.41.7.20250123_102326.zos/lib/5.41.7//os390
    /home/itodoro/zopen/usr/local/zopen/perl/perl5-heads.v5.41.7.20250123_102326.zos/lib/5.41.7/
    /home/itodoro/zopen/usr/local/zopen/perl/perl5-heads.v5.41.7.20250123_102326.zos/lib/5.41.7/os390/
    /home/itodoro/zopen/usr/local/zopen/perl/perl5-heads.v5.41.7.20250123_102326.zos/lib/perl5/5.41.7/
    /home/itodoro/zopen/usr/local/zopen/perl/perl5-heads.v5.41.7.20250123_102326.zos/lib/perl5/5.41.7/os390/
    /home/itodoro/zopen/usr/local/zopen/perl/perl5-heads.v5.41.7.20250123_102326.zos/lib/5.41.7//os390
    /home/itodoro/zopen/usr/local/zopen/perl/perl5-heads.v5.41.7.20250123_102326.zos/lib/5.41.7/
    /home/itodoro/zopen/usr/local/zopen/perl/perl5-heads.v5.41.7.20250123_102326.zos/lib/5.41.7/os390/
    /home/itodoro/zopen/usr/local/zopen/perl/perl5-heads.v5.41.7.20250123_102326.zos/lib/perl5/5.41.7/
    /home/itodoro/zopen/usr/local/zopen/perl/perl5-heads.v5.41.7.20250123_102326.zos/lib/perl5/5.41.7/os390/
    ../lib/site_perl/5.41.7/os390
    ../lib/site_perl/5.41.7
    ../lib/5.41.7/os390
    ../lib/5.41.7

```

Curl fails at `make check` with Perl 5.41.8 due to this as well:
```
make[2]: Leaving directory '/jenkins/workspace/Port-Build/curl/curl-8.11.1/tests'
srcdir=. /jenkins/zopen/usr/local/zopen/perl/perl5-heads.v5.41.7.20250123_102326.zos/bin/perl -I. ./runtests.pl   -a -s 
Possible precedence problem between ! and string eq at ./runtests.pl line 1691.
make[1]: *** [Makefile:898: quiet-test] Error 129
make[1]: Leaving directory '/jenkins/workspace/Port-Build/curl/curl-8.11.1/tests'
```

--- a/op.c
+++ b/op.c
@@ -12302,6 +12302,10 @@ check_precedence_not_vs_cmp(pTHX_ const OP *const o)
             && right->op_type == OP_CONST
             && SvIsBOOL(cSVOPx_sv(right))
         )
+        && (                                /* ... nor  !!$x == ...      */
+            (cUNOPx(left)->op_first->op_flags & OPf_PARENS)
+            || cUNOPx(left)->op_first->op_type != OP_NOT
+        )
     ) {
         Perl_ck_warner(aTHX_ packWARN(WARN_PRECEDENCE),
             "Possible precedence problem between ! and %s", OP_DESC(o)

--- a/pod/perldelta.pod
+++ b/pod/perldelta.pod
@@ -230,6 +230,14 @@ XXX Changes (i.e. rewording) of diagnostic messages go here
 
 =item *
 
+L<Possible precedence problem between ! and %s|perldiag/"Possible precedence problem between ! and %s">
+
+This warning no longer triggers for code like C<!!$x == $y>, i.e. where double
+negation (C<!!>) is used as a convert-to-boolean operator.
+[GH #22954]
+
+=item *
+
 XXX Describe change here
 
 =back

--- a/t/lib/warnings/op
+++ b/t/lib/warnings/op
@@ -1590,6 +1590,8 @@ for my $op (qw( == != < > <= >= eq ne lt gt le ge )) {
     $code .= "\$a = !\$a $op \$b;\n";  # should warn
     $code .= "\$a = (!\$a) $op \$b;\n";
     $code .= "\$a = !(\$a) $op \$b;\n";  # should warn
+    $code .= "\$a = !!\$a $op \$b;\n";
+    $code .= "\$a = !(!\$a) $op \$b;\n";  # should warn
     $code .= "\$a = !(\$a $op \$b);\n";
     $code .= "\$a = !\$a $op !0;\n";
     $code .= "\$a = !\$a $op !A_CONSTANT;\n";
@@ -1623,28 +1625,40 @@ Possible precedence problem between ! and derived class test at - line 6.
 Possible precedence problem between ! and derived class test at - line 8.
 Possible precedence problem between ! and numeric eq (==) at [cmp] line 1.
 Possible precedence problem between ! and numeric eq (==) at [cmp] line 3.
-Possible precedence problem between ! and numeric ne (!=) at [cmp] line 8.
+Possible precedence problem between ! and numeric eq (==) at [cmp] line 5.
 Possible precedence problem between ! and numeric ne (!=) at [cmp] line 10.
-Possible precedence problem between ! and numeric lt (<) at [cmp] line 15.
-Possible precedence problem between ! and numeric lt (<) at [cmp] line 17.
-Possible precedence problem between ! and numeric gt (>) at [cmp] line 22.
-Possible precedence problem between ! and numeric gt (>) at [cmp] line 24.
-Possible precedence problem between ! and numeric le (<=) at [cmp] line 29.
-Possible precedence problem between ! and numeric le (<=) at [cmp] line 31.
-Possible precedence problem between ! and numeric ge (>=) at [cmp] line 36.
-Possible precedence problem between ! and numeric ge (>=) at [cmp] line 38.
-Possible precedence problem between ! and string eq at [cmp] line 43.
-Possible precedence problem between ! and string eq at [cmp] line 45.
-Possible precedence problem between ! and string ne at [cmp] line 50.
-Possible precedence problem between ! and string ne at [cmp] line 52.
-Possible precedence problem between ! and string lt at [cmp] line 57.
-Possible precedence problem between ! and string lt at [cmp] line 59.
-Possible precedence problem between ! and string gt at [cmp] line 64.
-Possible precedence problem between ! and string gt at [cmp] line 66.
-Possible precedence problem between ! and string le at [cmp] line 71.
-Possible precedence problem between ! and string le at [cmp] line 73.
-Possible precedence problem between ! and string ge at [cmp] line 78.
-Possible precedence problem between ! and string ge at [cmp] line 80.
+Possible precedence problem between ! and numeric ne (!=) at [cmp] line 12.
+Possible precedence problem between ! and numeric ne (!=) at [cmp] line 14.
+Possible precedence problem between ! and numeric lt (<) at [cmp] line 19.
+Possible precedence problem between ! and numeric lt (<) at [cmp] line 21.
+Possible precedence problem between ! and numeric lt (<) at [cmp] line 23.
+Possible precedence problem between ! and numeric gt (>) at [cmp] line 28.
+Possible precedence problem between ! and numeric gt (>) at [cmp] line 30.
+Possible precedence problem between ! and numeric gt (>) at [cmp] line 32.
+Possible precedence problem between ! and numeric le (<=) at [cmp] line 37.
+Possible precedence problem between ! and numeric le (<=) at [cmp] line 39.
+Possible precedence problem between ! and numeric le (<=) at [cmp] line 41.
+Possible precedence problem between ! and numeric ge (>=) at [cmp] line 46.
+Possible precedence problem between ! and numeric ge (>=) at [cmp] line 48.
+Possible precedence problem between ! and numeric ge (>=) at [cmp] line 50.
+Possible precedence problem between ! and string eq at [cmp] line 55.
+Possible precedence problem between ! and string eq at [cmp] line 57.
+Possible precedence problem between ! and string eq at [cmp] line 59.
+Possible precedence problem between ! and string ne at [cmp] line 64.
+Possible precedence problem between ! and string ne at [cmp] line 66.
+Possible precedence problem between ! and string ne at [cmp] line 68.
+Possible precedence problem between ! and string lt at [cmp] line 73.
+Possible precedence problem between ! and string lt at [cmp] line 75.
+Possible precedence problem between ! and string lt at [cmp] line 77.
+Possible precedence problem between ! and string gt at [cmp] line 82.
+Possible precedence problem between ! and string gt at [cmp] line 84.
+Possible precedence problem between ! and string gt at [cmp] line 86.
+Possible precedence problem between ! and string le at [cmp] line 91.
+Possible precedence problem between ! and string le at [cmp] line 93.
+Possible precedence problem between ! and string le at [cmp] line 95.
+Possible precedence problem between ! and string ge at [cmp] line 100.
+Possible precedence problem between ! and string ge at [cmp] line 102.
+Possible precedence problem between ! and string ge at [cmp] line 104.
 Possible precedence problem between ! and pattern match (m//) at [bind] line 1.
 Possible precedence problem between ! and pattern match (m//) at [bind] line 3.
 Possible precedence problem between ! and pattern match (m//) at [bind] line 6.

